weather_mv/loader_pipeline/ee.py

```diff
@@ -0,0 +1,371 @@
+            )
+        else:
+            (
+                paths
+                | 'Log Grib Files' >> beam.Map(logger.debug)
+            )
+
+
+@dataclasses.dataclass
+class ConvertToCog(beam.DoFn):
+    """Writes tiff image after extracting grib data and uploads it to GCS.
+
+    Attributes:
+        tiff_location: The bucket location at which tiff files will be pushed.
+        open_dataset_kwargs: A dictionary of kwargs to pass to xr.open_dataset().
+        disable_in_memory_copy: A flag to turn in-memory copy off; Default: on.
+        disable_grib_schema_normalization: A flag to turn grib schema normalization off; Default: on.
+    """
+
+    tiff_location: str
+    open_dataset_kwargs: t.Optional[t.Dict] = None
+    disable_in_memory_copy: bool = False
+    disable_grib_schema_normalization: bool = False
+
+    def _get_tiff_name(self, uri: str) -> str:
+        """Extracts file name and converts it into an EE-safe name"""
+        file_name = os.path.basename(uri)
+        base_name = os.path.splitext(file_name)[0]
+        tiff_name = base_name.replace('.', '_')
+        return tiff_name
+
+    def _get_target_path(self, file_name: str) -> str:
+        """Creates a GCS target path from the file name."""
+        return os.path.join(self.tiff_location, file_name)
+
+    def process(self, uri: str) -> t.Iterator[TiffData]:
+        """Opens grib files and yields TiffData."""
+
+        logger.info(f'Converting {uri!r} to COGs ...')
+        with open_dataset(uri,
+                          self.open_dataset_kwargs,
+                          self.disable_in_memory_copy,
+                          self.disable_grib_schema_normalization) as ds:
+
+            attrs = ds.attrs
+            data = list(ds.values())
+            tiff_name = self._get_tiff_name(uri)
+            start_time, end_time = (attrs.get(key) for key in ('start_time', 'end_time'))
+            dtype, crs, transform = (attrs.pop(key) for key in ['dtype', 'crs', 'transform'])
+
+            file_name = f'{tiff_name}.tiff'
+
+            with tempfile.TemporaryDirectory() as temp_dir:
+                local_tiff_path = os.path.join(temp_dir, file_name)
+                with rasterio.open(local_tiff_path, 'w',
+                                   driver='GTiff',
+                                   dtype=dtype,
+                                   width=data[0].data.shape[1],
+                                   height=data[0].data.shape[0],
+                                   count=len(data),
+                                   crs=crs,
+                                   transform=transform,
+                                   tiled=True) as f:
+                    for i, da in enumerate(data):
+                        f.write(da, i+1)
+
+                # Copy local tiff to gcs.
+                target_path = self._get_target_path(file_name)
+                with open(local_tiff_path, 'rb') as src:
```

Comment:
I'm a little confused by this part. Can't you just have rasterio open up an in-memory buffer and then pass that buffer directly to the shutil.copyfileobj call? This seems like a disk write that you don't have to do...

Reply:
Agreed. Updated it.